File: 2448-count-number-of-bad-pairs/2448-count-number-of-bad-pairs.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public long countBadPairs(int[] arr) {
        int n = arr.length;
        Map<Integer, Integer> mp = new HashMap<>();
        long total = (long) n * (n - 1) / 2; // Total number of pairs
        long good = 0L;

        // Count occurrences of (i - arr[i]) and count good pairs directly
        for (int i = 0; i < n; i++) {
            int key = i - arr[i];
            good += mp.getOrDefault(key, 0); 
            mp.put(key, mp.getOrDefault(key, 0) + 1);
        }

        // Bad pairs = Total pairs - Good pairs
        return total - good;
    }
}

File: 2448-count-number-of-bad-pairs/2448-count-number-of-bad-pairs.cpp
class Solution {
public:
    long long countBadPairs(vector<int>& arr) {
        int n = arr.size();
        long long total = (long long)n * (n - 1) / 2;
        long long good=0l;
        unordered_map<int, int> mp;

        for (int i = 0; i < n; i++) {
            int key = i - arr[i];
            if (mp.find(key)!=mp.end())
                good += mp[key];
            mp[key]++;
        }

        return total - good;
    }
};